lib/entry-index.js

```diff
@@ -66,79 +86,58 @@ function find (cache, key) {
   })
 }
 
-module.exports.delete = del
 function del (cache, key) {
   return insert(cache, key, null)
 }
 
-module.exports.lsStream = lsStream
 function lsStream (cache) {
   const indexDir = bucketDir(cache)
-  const stream = through.obj()
-  fs.readdir(indexDir, function (err, buckets) {
-    if (err && err.code === 'ENOENT') {
-      return stream.end()
-    } else if (err) {
-      return stream.emit('error', err)
-    } else {
-      asyncMap(buckets, (bucket, cb) => {
-        fs.readdir(path.join(indexDir, bucket), (err, files) => {
-          if (err && err.code === 'ENOENT') {
-            return cb()
-          } else if (err) {
-            return cb(err)
-          } else {
-            asyncMap(files, function (f, cb) {
-              const bpath = path.join(indexDir, bucket, f)
-              bucketEntries(cache, bpath).then(_entries => {
-                const entries = _entries.reduce((acc, entry) => {
-                  acc[entry.key] = entry
-                  return acc
-                }, {})
-                Object.keys(entries).forEach(function (k) {
-                  stream.write(formatEntry(cache, entries[k]))
-                })
-                cb()
-              }, err => {
-                if (err.code === 'ENOENT') {
-                  cb()
-                } else {
-                  cb(err)
-                }
-              })
-            }, cb)
-          }
-        })
-      }, function (err) {
-        if (err) { stream.emit('error') }
-        stream.end()
-      })
-    }
+  const stream = from.obj()
+
+  // "/cachename/*"
+  readdirAsync(indexDir).map(bucket => {
+    const bucketPath = path.join(indexDir, bucket)
+
+    // "/cachename/<bucket 0xFF>/*"
+    return readdirAsync(bucketPath).map(subbucket => {
+      const subbucketPath = path.join(bucketPath, subbucket)
+
+      // "/cachename/<bucket 0xFF>/<bucket 0xFF>/*"
+      return readdirAsync(subbucketPath).map(entry => {
+        const getKeyToEntry = bucketEntries(
+          cache,
+          path.join(subbucketPath, entry)
+        ).reduce((acc, entry) => {
+          acc.set(entry.key, entry)
+          return acc
+        }, new Map())
+
+        return getKeyToEntry.then(reduced => {
+          return Array.from(reduced.values()).map(
+            entry => stream.push(formatEntry(cache, entry))
+          )
+        }).catch({code: 'ENOENT'}, nop)
+      }).catch({code: 'ENOENT'}, nop)
+    }).catch({code: 'ENOENT'}, nop)
+  }).catch({code: 'ENOENT'}, nop).then(() => {
```

Comment:
@iarna it can skip individual directory/file failures without failing the entire ls operation, which goes along with our "well if it's shit just ignore it" strategy for reading the cache.